### PR TITLE
[6174] Feature/ui log viewer downloadbtn

### PR DESCRIPTION
**Pull request by PGuimarais**
- Enables download button
- Names downloaded files according to CDAP program
- Changes timeline labels from 0-12hr to 0-24hr format to match table entries

Fixes Bugs:
- https://issues.cask.co/browse/CDAP-6776
- https://issues.cask.co/browse/CDAP-6693

### CODE REVIEW

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 1]
```diff
@@ -14,16 +14,19 @@
...
+function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_ACTIONS, MyCDAPDataSource, $sce, myCdapUrl, $timeout) {
   'ngInject';
 
   var dataSrc = new MyCDAPDataSource($scope);
   var pollPromise;
   //Collapsing LogViewer Table Columns
   var columnsList = [];
   var collapseCount = 0;
+  //If we are replacing a previously generated file we need to manuually revoke the object URL to avoid memory leaks
+  this.textFile = null;
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: I don't think we need this here. We are calling `setDefault()` at line 69 and that should set `this.textFile = null`

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 2]
```diff
@@ -270,6 +273,55 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
     });
   };
 
+  var exportTimeout = null;
+
+  const downloadLogs = () => {
+    return myLogsApi.getLogsStartAsRaw({
+          'namespace' : this.namespaceId,
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: I think the indentation is off here.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 3]
```diff
@@ -270,6 +273,55 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+          'programId' : this.programId,
+          'runId' : this.runId,
+          'start' : this.startTimeSec
+    }).$promise.then(
+      (res) => {
+        this.downloadContent = res;
+      },
+      (err) => {
+        console.log('ERROR: ', err);
+        return;
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Don't think we need a return statement. We should surface errors if any while fetching the logs for download. User should get a feedback if something goes wrong.

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 4]
```diff
@@ -270,6 +273,55 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+      var blob = new Blob([this.downloadContent], {type: 'text/plain'});
+        $scope.url = URL.createObjectURL(blob);
+        let filename = '';
+        if ('undefined' !== typeof this.getDownloadFilename()) {
+          filename = this.getDownloadFilename();
+        } else {
+          filename = this.namespaceId + '-' + this.appId + '-' + this.programId + '-' + this.startTimeSec;
+        }
+        $scope.exportFileName = filename;
+        $scope.$on('$destroy', () => {
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js:305]: I think we should consolidate all `$scope.$on('$destroy', () => {...}`. There is one here and one in line 321

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 5]
```diff
@@ -270,6 +271,49 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+  var exportTimeout = null;
+
+  const downloadLogs = () => {
+    return myLogsApi.getLogsStartAsRaw({
+      'namespace' : this.namespaceId,
+      'appId' : this.appId,
+      'programType' : this.programType,
+      'programId' : this.programId,
+      'runId' : this.runId,
+      'start' : this.startTimeSec
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: these keys do not need to be a string

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 6]
```diff
@@ -270,6 +271,49 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+    },
+    (err) => {
+      console.log('ERROR: ', err);
+    });
+  };
+
+  this.export = () => {
+    downloadLogs().then( () => {
+      var blob = new Blob([this.downloadContent], {type: 'text/plain'});
+        $scope.url = URL.createObjectURL(blob);
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: Any particular reason why attaching `url` to `$scope`?

File: cdap-ui/app/features/flows/templates/tabs/runs/tabs/log.html [hunk 7]
```diff
@@ -14,12 +14,14 @@
...
 -->
 
+
 <div ng-controller="FlowsRunDetailLogController as LogsController">
   <my-log-viewer
     namespace-id="{{LogsController.params.namespaceId}}"
     app-id="{{LogsController.params.appId}}"
     program-type="{{LogsController.params.programType}}"
     program-id="{{LogsController.params.programId}}"
+    get-download-filename="LogsController.getDownloadFileName()"
```
Comment (elfenheart) [inline cdap-ui/app/features/flows/templates/tabs/runs/tabs/log.html]: where is this function defined?

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 8]
```diff
@@ -270,6 +271,49 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+    downloadLogs().then( () => {
+      var blob = new Blob([this.downloadContent], {type: 'text/plain'});
+        this.url = URL.createObjectURL(blob);
+        let filename = '';
+        if ('undefined' !== typeof this.getDownloadFilename()) {
+          filename = this.getDownloadFilename();
+        } else {
+          filename = this.namespaceId + '-' + this.appId + '-' + this.programId + '-' + this.startTimeSec;
+        }
+        $scope.exportFileName = filename;
```
Comment (elfenheart) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: one more left over `$scope` variable

File: cdap-ui/app/directives/log-viewer/log-viewer.js [hunk 9]
```diff
@@ -270,6 +271,49 @@ function LogViewerController ($scope, LogViewerStore, myLogsApi, LOGVIEWERSTORE_
...
+
+  this.export = () => {
+    downloadLogs().then( () => {
+      var blob = new Blob([this.downloadContent], {type: 'text/plain'});
+        this.url = URL.createObjectURL(blob);
+        let filename = '';
+        if ('undefined' !== typeof this.getDownloadFilename()) {
+          filename = this.getDownloadFilename() + '-' + this.startTimeSec;
+        } else {
+          filename = this.namespaceId + '-' + this.appId + '-' + this.programId + '-' + this.startTimeSec;
```
Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: I believe this should have runId included. @bdmogal do we want run id in the file name while downloading them?

Comment (bdmogal) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: I had discussed with @PGuimarais. The runids we have are not very user-friendly. So I thought we should include start time instead. @nitinmotgi, @ToddGreenstein thoughts?

Comment (ajainarayanan) [inline cdap-ui/app/directives/log-viewer/log-viewer.js]: The star time is in epoch format. I am not sure if that will be user friendly though..

File: cdap-ui/app/features/services/controllers/tabs/runs/tabs/log-ctrl.js [hunk 10]
```diff
@@ -21,6 +21,10 @@ angular.module(PKG.name + '.feature.services')
       return;
     }
 
+    this.getDownloadFileName = () => {
```
Comment (elfenheart) [inline cdap-ui/app/features/services/controllers/tabs/runs/tabs/log-ctrl.js]: why only for services?

Comment (PGuimarais) [inline cdap-ui/app/features/services/controllers/tabs/runs/tabs/log-ctrl.js]: CDAP apps follow the same naming convention. Hydrator and Tracker will also receive their custom naming convention. @ajainarayanan will cover Hydrator in his PR, and I will cover Tracker once it has been integrated.

Comment (elfenheart) [inline cdap-ui/app/features/services/controllers/tabs/runs/tabs/log-ctrl.js]: if that's the case, then it is not needed here, right?

File: cdap-ui/app/features/services/templates/tabs/runs/tabs/log.html [hunk 11]
```diff
@@ -20,6 +20,7 @@
       app-id="{{LogsController.params.appId}}"
       program-type="{{LogsController.params.programType}}"
       program-id="{{LogsController.params.programId}}"
-      run-id="{{LogsController.params.runId}}">
+      run-id="{{LogsController.params.runId}}"
+      get-download-filename="LogsController.getDownloadFileName()">
```
Comment (elfenheart) [inline cdap-ui/app/features/services/templates/tabs/runs/tabs/log.html]: should remove from here too

**Comment by elfenheart** [comment]
👍  LGTM

**Pull request merged**